ledger/acctupdates.go

```diff
@@ -1701,12 +1725,60 @@ func (au *accountUpdates) getCreatorForRoundImpl(rnd basics.Round, cidx basics.C
 				au.accountsReadCond.Wait()
 			}
 		} else {
-			au.log.Errorf("accountUpdates.getCreatorForRoundImpl: database round %d mismatching in-memory round %d", dbRound, currentDbRound)
+			au.log.Errorf("accountUpdates.getCreatorForRound: database round %d mismatching in-memory round %d", dbRound, currentDbRound)
 			return basics.Address{}, false, &MismatchingDatabaseRoundError{databaseRound: dbRound, memoryRound: currentDbRound}
 		}
 	}
 }
 
+func (au *accountUpdates) countStorageForRound(rnd basics.Round, addr basics.Address, aidx basics.AppIndex, global bool) (basics.StateSchema, error) {
+	ad, _, err := au.lookupWithoutRewards(rnd, addr, false /* do not take lock */)
+	if err != nil {
+		return basics.StateSchema{}, err
+	}
+
+	count := basics.StateSchema{}
+	kv := basics.TealKeyValue{}
+	if global {
+		if app, ok := ad.AppParams[aidx]; ok {
+			kv = app.GlobalState
+		}
+	} else {
+		if ls, ok := ad.AppLocalStates[aidx]; ok {
+			kv = ls.KeyValue
+		}
+	}
+
+	for _, v := range kv {
+		if v.Type == basics.TealUintType {
+			count.NumUint++
+		} else {
+			count.NumByteSlice++
+		}
+	}
+	return count, nil
+}
+
+func (au *accountUpdates) getKeyForRound(rnd basics.Round, addr basics.Address, aidx basics.AppIndex, global bool, key string) (basics.TealValue, bool, error) {
```

Comment:
the implementation doesn't seem to return a different value in the case where the account doesn't have that application vs. the account does have the application, but doesn't have that particular key.